### PR TITLE
Add doOn shortcut operators

### PR DESCRIPTION
**Pull request by thanegill**
Adds three operators that are shortcuts for `doOn`. I use these commonly and have seem them in [other projects](https://github.com/artsy/eidolon/blob/04a0c8488525be4b918e212e92a29f143e0f5838/Kiosk/Observable%2BOperators.swift#L64-L94) as well.

### CODE REVIEW

**Comment by ashfurrow** [comment]
Ooohhhhhhhhh I like this a lot! 💯

**Comment by alcarvalho** [comment]
:+1: for this!

**Comment by tailec** [comment]
:+1:

**Comment by ashfurrow** [comment]
I think this looks good, and while I have rights to merge I'm wary of doing so without a :+1: from @bontoJR or @kzaher. I imagine this could use tests/changelog update in the meantime. Looks great though! Documentation is :tada:

**Comment by bontoJR** [comment]
Loved this shortcut!

Now let 's wait for @kzaher to decide if merge it or not. I am definitely in with this!

**Comment by kzaher** [comment]
Hi guys,

I guess if you are like it so much, I'm fine with merging this. Could we just make a single unit test per extension to make sure this doesn't break in future before we can merge this?

**Comment by thanegill** [comment]
@kzaher I don't see any tests for `doOn(onNext:onError:onCompleted:)`, should those be added as well? Also should I migrate any examples to use these new shortcuts?

**Comment by kzaher** [comment]
@thanegill If you could add some tests for `doOn(onNext:onError:onCompleted)`, what would be great.
We could probably add one for `onNext`, `onError` and `onCompleted`.

We would also want to test that if there is error thrown in `do*` closure, that sequence is terminated with that error. So that's additional 6 tests. (3 for shortcuts, and 3 for original operator).

If you can add them that would be :+1:

If you feel we should migrate some code to use those new shortcuts, feel free to migrate it. I'm only aware of one place in example app that uses them.

**Comment by thanegill** [comment]
@kzaher Yeah I can add the tests. I might take me a little while to grok how to implement the tests. If there's a good example to start from that would be great.

I only saw `doOn(onNext:)` used in `ImageService` I'll migrate those to `doOnNext()`.

**Comment by thanegill** [comment]
@kzaher There are tests for `doOn(onNext:onError:onCompleted:)` There names just didn't transition from `do` -> `doOn`.

**Comment by kzaher** [comment]
Haha, when you've said we didn't have them, sounded plausible to me :) 
Didn't check your claim.

You are correct, those are the tests. I should have just searched for `doOn` :)

I think that we would then need 6 tests. 3 for normal path, and 3 for thrown errors.

For thrown errors we could probably do something like

```
    func testDoOnError_throws() {
        let scheduler = TestScheduler(initialClock: 0)
        
        let xs = scheduler.createHotObservable([
            next(150, 1),
            next(210, 2),
            error(250, testError)
            ])
        
        let res = scheduler.start { xs.doOnError { _ in
                throw testError1
            }
        }
        
        let correctMessages = [
             next(210, 2),
             error(250, testError1)
        ]
        
        let correctSubscriptions = [
            Subscription(200, 250)
        ]
        
        XCTAssertEqual(res.events, correctMessages)
        XCTAssertEqual(xs.subscriptions, correctSubscriptions)
    }
```

For normal case, something like 

```
    func testDoOnError_normal() {
        let scheduler = TestScheduler(initialClock: 0)
        
        let xs = scheduler.createHotObservable([
            next(150, 1),
            next(210, 2),
            error(250, testError)
            ])
        
        var recordedError: ErrorType!
        var numberOfTimesInvoked = 0
        let res = scheduler.start { xs.doOnError { error in
                recordedError = error
                numberOfTimesInvoked = numberOfTimesInvoked + 1
            }
        }
        
        let correctMessages = [
             next(210, 2),
             error(250, testError)
        ]
        
        let correctSubscriptions = [
            Subscription(200, 250)
        ]
        
        XCTAssertEqual(res.events, correctMessages)
        XCTAssertEqual(xs.subscriptions, correctSubscriptions)
        
        XCTAssertEqual(recordedError, testError)
        XCTAssertEqual(numberOfTimesInvoked, 1)
    }
```

We can also rename old unit tests to be more clear which operator is being tested.

**Comment by thanegill** [comment]
@kzaher Test's have been implemented. I can squash the commit's if you'd like.

**Comment by thanegill** [comment]
Don't merge yet. I'm adding them to Driver as well.

**Comment by thanegill** [comment]
@kzaher Right now the `testAsDriver_doOnError` test fails. Should Driver even have a `doOnError` operator?

**Comment by kzaher** [comment]
@thanegill driver can't fail otherwise we are dooomed :)
doomed

**Comment by thanegill** [comment]
@kzaher Thought so. I removed `onError` from `doOn` as well.

File: RxSwift/Observables/Observable+Single.swift [hunk 1]
```diff
@@ -113,6 +113,42 @@ extension ObservableType {
...
+    }
+
+    /**
+     Invokes an action for the Completed event in the observable sequence, and propagates all observer messages through the result sequence.
+
+     - parameter onCompleted: Action to invoke upon graceful termination of the observable sequence.
+     - returns: The source sequence with the side-effecting behavior applied.
+     */
+    @warn_unused_result(message="http://git.io/rxs.uo")
+    public func doOnComplete(onCompleted: (() throws -> Void))
```
Comment (kzaher) [inline RxSwift/Observables/Observable+Single.swift]: I think this should be named `doOnCompleted`.

File: RxCocoa/Common/CocoaUnits/Driver/Driver+Operators.swift [hunk 2]
```diff
@@ -129,13 +129,37 @@ extension DriverConvertibleType {
...
+    }
+
+    /**
+     Invokes an action for the Completed event in the observable sequence, and propagates all observer messages through the result sequence.
+
+     - parameter onCompleted: Action to invoke upon graceful termination of the observable sequence.
+     - returns: The source sequence with the side-effecting behavior applied.
+     */
+    @warn_unused_result(message="http://git.io/rxs.uo")
+    public func doOnComplete(onCompleted: (() -> Void))
```
Comment (kzaher) [inline RxCocoa/Common/CocoaUnits/Driver/Driver+Operators.swift]: I believe this one too.

File: Tests/RxSwiftTests/Tests/Observable+SingleTest.swift [hunk 3]
```diff
@@ -558,6 +558,213 @@ extension ObservableSingleTest {
...
+        let scheduler = TestScheduler(initialClock: 0)
+
+        let xs = scheduler.createHotObservable([
+            next(150, 1),
+            next(210, 2),
+            error(250, testError)
+            ])
+
+        let res = scheduler.start { xs.doOnError { _ in
+                throw testError
```
Comment (kzaher) [inline Tests/RxSwiftTests/Tests/Observable+SingleTest.swift]: Could you maybe throw `testError1` here and then have
```
let correctMessages = [
 +            next(210, 2),
 +            error(250, testError1)
 +        ]
```

... so we know that error thrown in `doOnError` overrides original error.

**Comment by kzaher** [comment]
A couple of small fixes, and we are ready to pull this in :)

**Comment by thanegill** [comment]
Done and done 🎉

**Pull request merged**